File: ios-native/Sources/PolarisStudyKit/PolarisStudy.swift
import Foundation

/// PolarisStudy — the on-device study toolkit, powered entirely by **classical algorithms,
/// not AI**. Every feature is deterministic, offline, instant, and testable:
///
/// | Area        | Algorithm |
/// |-------------|-----------|
/// | Flashcards  | SM-2 spaced repetition + cloze/definition generation |
/// | Quizzing    | MCQ generation + Levenshtein grading + Leitner boxes |
/// | Citation    | Rule-table formatter (APA / MLA / Chicago) |
/// | Essay       | Flesch–Kincaid readability + structural heuristics |
/// | CV Builder  | Template rendering |
/// | Advisor     | Rule-based, deadline + spacing study scheduler |
///
/// No model download, no network, no `async`, no Apple-Intelligence requirement.
public struct PolarisStudy: Sendable {

    public init() {}

    // MARK: Flashcards (SM-2)
    public func makeFlashcards(from text: String, limit: Int = 20) -> [Flashcard] {
        FlashcardGenerator.generate(from: text, limit: limit)
    }

    public func review(_ state: SM2State, grade: Int) -> SM2State {
        SpacedRepetition.schedule(state, grade: grade)
    }

    // MARK: Quizzing
    public func makeQuiz(from cards: [Flashcard], choices: Int = 4) -> [QuizQuestion] {
        var rng = SystemRandomNumberGenerator()
        return QuizEngine.multipleChoice(from: cards, choices: choices, rng: &rng)
    }

    public func grade(_ given: String, answer: String) -> Bool {
        QuizEngine.isCorrect(given, answer: answer)
    }

    // MARK: Citation
    public func citation(_ source: Source, style: CitationStyle) -> (reference: String, inText: String) {
        (Citations.reference(source, style: style), Citations.inText(source, style: style))
    }

    // MARK: Essay
    public func analyzeWriting(_ text: String) -> ReadabilityReport {
        Readability.analyze(text)
    }

    // MARK: CV Builder (template)
    public struct Resume: Sendable {
        public var name = "", contact = "", summary = ""
        public var experience: [String] = [], education: [String] = [], skills: [String] = []
        public init() {}
    }

    /// Render a résumé as Markdown from structured fields — pure templating, no AI.
    public func buildCV(_ r: Resume) -> String {
        var out = ["# \(r.name)"]
        if !r.contact.isEmpty { out.append(r.contact) }
        if !r.summary.isEmpty { out += ["", "## Summary", r.summary] }
        func section(_ title: String, _ items: [String]) {
            guard !items.isEmpty else { return }
            out += ["", "## \(title)"] + items.map { "- \($0)" }
        }
        section("Experience", r.experience)
        section("Education", r.education)
        if !r.skills.isEmpty { out += ["", "## Skills", r.skills.joined(separator: " · ")] }
        return out.joined(separator: "\n")
    }

    // MARK: Advisor (rule-based scheduler)
    public struct Deadline: Sendable {
        public let title: String
        public let due: Date
        public let weightPct: Double
        public init(title: String, due: Date, weightPct: Double = 0) {
            self.title = title; self.due = due; self.weightPct = weightPct
        }
    }

    public struct PlanItem: Sendable {
        public let title: String
        public let start: Date
        public let priority: Double
    }

    /// Prioritize upcoming work by urgency × importance — no AI.
    /// priority = weight / max(1, days-until-due); suggested start backs off from the deadline
    /// by more days for heavier items (front-loading).
    public func advise(_ deadlines: [Deadline], now: Date = Date()) -> [PlanItem] {
        let cal = Calendar.current
        return deadlines.map { d in
            let days = max(1, cal.dateComponents([.day], from: now, to: d.due).day ?? 1)
            let priority = (d.weightPct > 0 ? d.weightPct : 10) / Double(days)
            let lead = min(days, Int((d.weightPct / 20).rounded()) + 1)   // heavier → start sooner
            let start = cal.date(byAdding: .day, value: -lead, to: d.due) ?? d.due
            return PlanItem(title: d.title, start: max(now, start), priority: priority)
        }
        .sorted { $0.priority > $1.priority }
    }
}

File: android-native/polaris-study-kit/src/main/java/com/polaris/studykit/PolarisStudy.kt
package com.polaris.studykit

import java.time.LocalDate
import java.time.temporal.ChronoUnit
import kotlin.math.roundToInt

/**
 * PolarisStudy — the on-device study toolkit, powered entirely by **classical algorithms,
 * not AI**. Every feature is deterministic, offline, instant, and unit-tested:
 *
 * | Area        | Algorithm |
 * |-------------|-----------|
 * | Flashcards  | SM-2 spaced repetition + cloze/definition generation |
 * | Quizzing    | MCQ generation + Levenshtein grading + Leitner boxes |
 * | Citation    | Rule-table formatter (APA / MLA / Chicago) |
 * | Essay       | Flesch–Kincaid readability + structural heuristics |
 * | CV Builder  | Template rendering |
 * | Advisor     | Rule-based deadline + spacing study scheduler |
 *
 * No model file, no MediaPipe, no network — nothing to download and it runs on any device.
 * All calls are cheap and synchronous (no [AutoCloseable], no background thread required).
 */
class PolarisStudy {

    // Flashcards (SM-2)
    fun makeFlashcards(text: String, limit: Int = 20): List<Flashcard> =
        FlashcardGenerator.generate(text, limit)

    fun review(state: Sm2State, grade: Int, today: Long = LocalDate.now().toEpochDay()): Sm2State =
        SpacedRepetition.schedule(state, grade, today)

    // Quizzing
    fun makeQuiz(cards: List<Flashcard>, choices: Int = 4): List<QuizQuestion> =
        QuizEngine.multipleChoice(cards, choices)

    fun grade(given: String, answer: String): Boolean = QuizEngine.isCorrect(given, answer)

    // Citation
    fun citation(source: Source, style: CitationStyle): Pair<String, String> =
        Citations.reference(source, style) to Citations.inText(source, style)

    // Essay
    fun analyzeWriting(text: String): ReadabilityReport = Readability.analyze(text)

    // Writing checker (offline). The AI coach "Polly" is ONLINE-ONLY — call its API
    // (/writing/coach, /writing/polish) only when the device has connectivity; this rule-based
    // checker always works, on-device, with no network.
    fun checkWriting(text: String): Pair<List<WritingIssue>, Int> {
        val issues = WritingChecker.check(text)
        return issues to WritingChecker.score(text, issues)
    }

    // CV Builder (template)
    data class Resume(
        val name: String = "",
        val contact: String = "",
        val summary: String = "",
        val experience: List<String> = emptyList(),
        val education: List<String> = emptyList(),
        val skills: List<String> = emptyList(),
    )

    fun buildCV(r: Resume): String {
        val out = mutableListOf("# ${r.name}")
        if (r.contact.isNotEmpty()) out.add(r.contact)
        if (r.summary.isNotEmpty()) out.addAll(listOf("", "## Summary", r.summary))
        fun section(title: String, items: List<String>) {
            if (items.isEmpty()) return
            out.add(""); out.add("## $title"); out.addAll(items.map { "- $it" })
        }
        section("Experience", r.experience)
        section("Education", r.education)
        if (r.skills.isNotEmpty()) out.addAll(listOf("", "## Skills", r.skills.joinToString(" · ")))
        return out.joinToString("\n")
    }

    // Advisor (rule-based scheduler)
    data class Deadline(val title: String, val due: LocalDate, val weightPct: Double = 0.0)
    data class PlanItem(val title: String, val start: LocalDate, val priority: Double)

    /**
     * Prioritize upcoming work by urgency × importance — no AI.
     * priority = weight / max(1, days-until-due); heavier items start sooner (front-loading).
     */
    fun advise(deadlines: List<Deadline>, now: LocalDate = LocalDate.now()): List<PlanItem> =
        deadlines.map { d ->
            val days = ChronoUnit.DAYS.between(now, d.due).coerceAtLeast(1)
            val priority = (if (d.weightPct > 0) d.weightPct else 10.0) / days
            val lead = ((d.weightPct / 20).roundToInt() + 1).toLong().coerceAtMost(days)
            val start = maxOf(now, d.due.minusDays(lead))
            PlanItem(d.title, start, priority)
        }.sortedByDescending { it.priority }
}
